### PR TITLE
Add optional uri_only serialization/deserialization support, logged planner, and replay-log script

### PR DESCRIPTION
**Pull request by cdellin**
This adds support for the `uri_only` flag to the PrPy serialization functions (defaults to `False`).  When this flag is set, the fixed aspects (e.g. geometry and kinematics) of kinbodies and robots in the scene that have URIs (e.g. OpenRAVE xml paths or URDF/SRDF file paths) are not serialized, and instead only the URI itself is serialized.  This is an immediate workaround for a bug in the serialization functions (or in OpenRAVE) #243 which precludes roundtripping HERB with the correct kinematics hashes, resulting in busted IK solvers.  Other advantages include reducing the size of the serialized files, and allowing planning requests to be replayed while robot or object models are modified.

This PR also contains a `Logged` planner, which takes a delegate planner, and logs both the incoming planning request and some simple data about the result, to a file.  There is also a corresponding `replay-log.py` script which allows the planning requests in these log files to be replayed.  (It is expected that downstream users, such ash HerbPy, may want to provide their own log replay facilities which can make use of robot-specific helpers and planners.)

We've been using this a bit for ISER, and it appears to be working well, so I think it's ready for a review.

### CODE REVIEW

File: src/prpy/serialization.py [hunk 1]
```diff
@@ -99,43 +101,60 @@ def serialize_environment_file(env, path, writer=None):
...
-def serialize_kinbody(body):
-    all_joints = []
-    all_joints.extend(body.GetJoints())
-    all_joints.extend(body.GetPassiveJoints())
+def serialize_kinbody(body, uri_only=False):
 
+    uri = body.GetXMLFilename()
+    
+    if not uri:
+        uri_only = False
```
Comment (mkoval) [inline src/prpy/serialization.py]: Print a warning in this case. It's odd that you can pass `uri_only`, but then have the implementation silently fall back on full serialization. Or add another option like `try_uri` that makes it clear that it's best-effort only.

File: src/prpy/serialization.py [hunk 2]
```diff
@@ -99,43 +101,60 @@ def serialize_environment_file(env, path, writer=None):
...
         'name': body.GetName(),
-        'uri': body.GetXMLFilename(),
-        'links': map(serialize_link, body.GetLinks()),
-        'joints': map(serialize_joint, all_joints),
+        'uri': uri,
     }
-    data['kinbody_state'] = serialize_kinbody_state(body)
+    
+    if uri_only:
+        data['uri_only'] = True
```
Comment (mkoval) [inline src/prpy/serialization.py]: Why not define `uri_only` in both cases, i.e. `data['uri_only'] = uri_only`?

File: src/prpy/serialization.py [hunk 3]
```diff
@@ -99,43 +101,60 @@ def serialize_environment_file(env, path, writer=None):
...
+def serialize_kinbody_state(body, uri_only=False):
     data = {
         name: get_fn(body)
         for name, (get_fn, _) in KINBODY_STATE_MAP.iteritems()
     }
 
     link_transforms, dof_branches = body.GetLinkTransformations(True)
     data.update({
-        'link_transforms': map(serialize_transform, link_transforms),
-        'dof_branches': dof_branches.tolist(),
```
Comment (mkoval) [inline src/prpy/serialization.py]: Why did you remove `dof_branches`?

File: src/prpy/serialization.py [hunk 4]
```diff
@@ -99,43 +101,60 @@ def serialize_environment_file(env, path, writer=None):
...
         'dof_values': body.GetDOFValues().tolist(),
     })
+    if not uri_only:
+        data.update({
+            'link_transforms': map(serialize_transform, link_transforms),
+        })
```
Comment (mkoval) [inline src/prpy/serialization.py]: `link_transformations` and `dof_branches` are needed to call `SetDOFValues` and get back bit-identical results. They should be serialized _even if_ `uri_only` is set.

File: src/prpy/serialization.py [hunk 5]
```diff
@@ -327,37 +344,66 @@ def deserialize_kinbody(env, data, name=None, anonymous=False, state=True):
...
-        kinbody = RaveCreateRobot(env, '')
-        kinbody.Init(
-            link_infos, joint_infos,
-            manipulator_infos, sensor_infos,
-            data['uri']
-        )
+    
+    name_desired = name or data['name']
+    
+    if 'uri_only' in data and data['uri_only']:
```
Comment (mkoval) [inline src/prpy/serialization.py]: Replace this with `data.get('uri_only', False)`.

File: src/prpy/serialization.py [hunk 6]
```diff
@@ -327,37 +344,66 @@ def deserialize_kinbody(env, data, name=None, anonymous=False, state=True):
...
+        
+        if data['is_robot']:
+            
+            if data['uri'].endswith('.robot.xml'):
+                kinbody = env.ReadRobotXMLFile('robots/barrettwam.robot.xml')
```
Comment (mkoval) [inline src/prpy/serialization.py]: Uhhhh..what?

File: src/prpy/serialization.py [hunk 7]
```diff
@@ -327,37 +344,66 @@ def deserialize_kinbody(env, data, name=None, anonymous=False, state=True):
...
+                kinbody.SetName(name_desired)
+                env.Add(kinbody, anonymous)
+            else:
```
Comment (mkoval) [inline src/prpy/serialization.py]: I would prefer to flip these two cases. OpenRAVE supports several file extensions (`.dae`, `.zae`, etc) that would not be correctly handled by this. I would prefer to _only_ use `or_urdf` if the file ends in `.urdf` or `.srdf`.

File: src/prpy/serialization.py [hunk 8]
```diff
@@ -327,37 +344,66 @@ def deserialize_kinbody(env, data, name=None, anonymous=False, state=True):
...
+                urdf,srdf = data['uri'].split()
```
Comment (mkoval) [inline src/prpy/serialization.py]: There is no point in `split()`ing this because you pass it directly to `or_urdf` a few lines below.

File: src/prpy/serialization.py [hunk 9]
```diff
@@ -327,37 +344,66 @@ def deserialize_kinbody(env, data, name=None, anonymous=False, state=True):
...
+                m_urdf = openravepy.RaveCreateModule(env, 'urdf')
```
Comment (mkoval) [inline src/prpy/serialization.py]: Is there any way we can create _one_ `or_urdf` module for the entire de-serialization, instead of creating a new one for each robot?

File: src/prpy/serialization.py [hunk 10]
```diff
@@ -327,37 +344,66 @@ def deserialize_kinbody(env, data, name=None, anonymous=False, state=True):
...
+                robot_name = m_urdf.SendCommand('Load {:s} {:s}'.format(urdf,srdf))
+                if robot_name != name_desired:
+                    raise RuntimeError('error, or_urdf name mismatch!')
+                kinbody = env.GetRobot(robot_name)
+            
+        else:
+            
+            kinbody = env.ReadKinBodyXMLFile(data['uri'])
+            kinbody.SetName(name_desired)
+            env.Add(kinbody)
```
Comment (mkoval) [inline src/prpy/serialization.py]: Why don't you handle `or_urdf` in this case?

File: src/prpy/serialization.py [hunk 11]
```diff
@@ -380,11 +426,14 @@ def deserialize_kinbody_state(body, data):
...
-        data['dof_branches']
-    )
+    
+    body.SetDOFValues(data['dof_values'])
+    
+    if 'link_transforms' in data:
+        body.SetLinkTransformations(
+            map(deserialize_transform, data['link_transforms']),
+            data['dof_branches']
+        )
```
Comment (mkoval) [inline src/prpy/serialization.py]: Please remove `SetDOFValues` and restore `SetLinkTransformations`: `SetDOFValues(GetDOFValues(...))` is not an identity-preserving operation.

**Comment by gilwoolee** [comment]
Thanks @cdellin ! It's been very helpful. As an extensive user to these additional features, I have one suggestion. As you already acknowledged in the PR comment, `replay-log.py`'s robot contains no information other than those from URDF, so for example it can't do `PlanToNamedConfiguration` because the robot does not have config_library. @Shushman was at some point confused about this too. I think it's confusing because it successfully loads the robot and is capable of handling some of the planning queries. :) 

There were a few other features I needed so I ended up just loading HERB from herbpy. I think this script serves its purpose as a base script to work from, but I think having an additional comment at the top of `replay-log.py` explaining that _downstream users may need to add some additional features or load robot based on the retrieved name_ may be useful.

File: src/prpy/serialization.py (hunk 1 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Fixed!

File: src/prpy/serialization.py (hunk 2 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Added a comment; this is so that old deserializers still work.

File: src/prpy/serialization.py (hunk 3 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Fixed! They were removed because I originally wanted to replicate just loading the robot and setting DOF values.

File: src/prpy/serialization.py (hunk 4 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Fixed.

File: src/prpy/serialization.py (hunk 5 above)
Comment (cdellin) [inline src/prpy/serialization.py]: I only slightly prefer my way, but I'm happy to change `(-:`.

File: src/prpy/serialization.py (hunk 6 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Ugh, I *guess* I can remove this.  But seriously, when was the last time that you wanted to load an XML file and it wasn't the WAM?  I think it's actually more likely that you wanted to load the WAM but got the path wrong (which this will fix!) than that you want to load some other robot. `(-:`

File: src/prpy/serialization.py (hunk 7 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Fixed.

File: src/prpy/serialization.py (hunk 8 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Fixed.

File: src/prpy/serialization.py (hunk 9 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Behold the `UnitaryMemoizer` -- it will fix all our problems~

File: src/prpy/serialization.py (hunk 10 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Added!

File: src/prpy/serialization.py (hunk 11 above)
Comment (cdellin) [inline src/prpy/serialization.py]: I added a case for both, so we can handle the ISER logs.

**Comment by cdellin** [comment]
Thanks @gilwoolee!  I added a comment to the top of `replay-log.py` to that effect.

**Comment by cdellin** [comment]
OK, I think everything's addressed.

File: src/prpy/planning/logged.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+import openravepy
+import numpy
+import yaml
+
+import prpy.planning.base
+import prpy.serialization
+
+
+class LoggedPlanner(prpy.planning.base.MetaPlanner):
```
Comment (mkoval) [inline src/prpy/planning/logged.py]: Please add a docstring.

File: src/prpy/planning/logged.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                reqdict['args'].append(prpy.serialization.serialize(arg))
+            except prpy.exceptions.UnsupportedTypeSerializationException as ex:
+                reqdict['args'].append(str(ex))
+        reqdict['kw_args'] = {}
+        for key,value in kw_args.items():
+            key = prpy.serialization.serialize(key)
+            try:
+                reqdict['kw_args'][key] = prpy.serialization.serialize(value)
+            except prpy.exceptions.UnsupportedTypeSerializationException as ex:
+                reqdict['kw_args'][key] = 'Error: ' + str(ex)
```
Comment (mkoval) [inline src/prpy/planning/logged.py]: Why are failures treated differently for `args` and `kwargs` (one is prefixed with `Error: `)? Also, I would prefer to factor this `try`/`except` block into a separate function to remove some copy-pasta.

File: src/prpy/planning/logged.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        reqdict['method'] = method # string
+        reqdict['args'] = []
+        for arg in args:
+            try:
+                reqdict['args'].append(prpy.serialization.serialize(arg))
+            except prpy.exceptions.UnsupportedTypeSerializationException as ex:
+                reqdict['args'].append(str(ex))
+        reqdict['kw_args'] = {}
+        for key,value in kw_args.items():
+            key = prpy.serialization.serialize(key)
```
Comment (mkoval) [inline src/prpy/planning/logged.py]: Missing error checking on the dictionary key.

File: src/prpy/planning/logged.py (hunk 13 above)
Comment (mkoval) [inline src/prpy/planning/logged.py]: You could even replace all of this with a wicked dictionary comprehension if you add that function! :grin:

File: src/prpy/planning/logged.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        
+        # get ready to serialize result
+        resdict = {}
+        
+        try:
+            plan_fn = getattr(self.planner, method)
+            traj = plan_fn(*args, **kw_args)
+            resdict['ok'] = True
+            resdict['traj_first'] = list(map(float,traj.GetWaypoint(0)))
+            resdict['traj_last'] = list(map(float,traj.GetWaypoint(traj.GetNumWaypoints()-1)))
```
Comment (mkoval) [inline src/prpy/planning/logged.py]: Why not serialize the entire trajectory?

File: src/prpy/serialization.py [hunk 16]
```diff
@@ -16,7 +16,9 @@ def serialize(obj):
 
     NoneType = type(None)
 
-    if isinstance(obj, (int, float, basestring, NoneType)):
+    if isinstance(obj, numpy.float64):
```
Comment (mkoval) [inline src/prpy/serialization.py]: Replace `numpy.float65` with `numpy.floating`, which is the base class for all NumPy floating point types (`float`, `float16`, `float32`, `float64`, and `float128`. You may want to add similar logic to convert `numpy.signedinteger` and `numpy.unsignedinteger` to `int`s.

File: src/prpy/serialization.py [hunk 17]
```diff
@@ -99,32 +101,51 @@ def serialize_environment_file(env, path, writer=None):
...
-    all_joints.extend(body.GetJoints())
-    all_joints.extend(body.GetPassiveJoints())
+def serialize_kinbody(body, uri_only=False):
 
+    uri = body.GetXMLFilename()
+    
+    if uri_only and not uri:
+        serialization_logger.warn(
+            'uri_only passed, but kinbody {} doesn\'t have a URI.'.format(
+            body.GetName()))
```
Comment (mkoval) [inline src/prpy/serialization.py]: Nit: `kinbody` -> `KinBody`. Also, I would prefer to specifically mention that `GetXMLFilename()` is empty. This will give the user a hint of where to look while debugging this.

File: src/prpy/serialization.py [hunk 18]
```diff
@@ -99,32 +101,51 @@ def serialize_environment_file(env, path, writer=None):
...
         'is_robot': body.IsRobot(),
         'name': body.GetName(),
-        'uri': body.GetXMLFilename(),
-        'links': map(serialize_link, body.GetLinks()),
-        'joints': map(serialize_joint, all_joints),
+        'uri': uri,
     }
-    data['kinbody_state'] = serialize_kinbody_state(body)
+
+    # only add uri_only to serialization so that older deserializers work
```
Comment (mkoval) [inline src/prpy/serialization.py]: Comment doesn't make sense. I think you mean:

> Only add `uri_only` to the serialization **if `uri_only` is set** so older deserializers work.

File: src/prpy/serialization.py [hunk 19]
```diff
@@ -99,32 +101,51 @@ def serialize_environment_file(env, path, writer=None):
...
         'robot_state': serialize_robot_state(robot),
     }
+    if not uri_only:
+        data.update({
+            'manipulators': map(serialize_manipulator, robot.GetManipulators()),
+        })
+    return data
 
-def serialize_kinbody_state(body):
+def serialize_kinbody_state(body, uri_only=False):
```
Comment (mkoval) [inline src/prpy/serialization.py]: This function no longer needs the `uri_only` parameter. It doesn't make sense conceptually, either, since the state must be serialized even if `uri_only` is `True`.

File: src/prpy/serialization.py [hunk 20]
```diff
@@ -195,6 +216,17 @@ def serialize_transform(t):
     }
 
 # Deserialization.
+
+class UnitaryMemoizer:
```
Comment (mkoval) [inline src/prpy/serialization.py]: Missing docstring.

File: src/prpy/serialization.py [hunk 21]
```diff
@@ -300,12 +330,16 @@ def deserialize_environment(data, env=None, purge=False, reuse_bodies=None):
             deserialization_logger.debug('Purging body "%s".', body.GetName())
             env.Remove(body)
 
+    # Create a or_ordf module on demand
+    urdf_module_getter = UnitaryMemoizer(lambda: openravepy.RaveCreateModule(env,'urdf'))
```
Comment (mkoval) [inline src/prpy/serialization.py]: Throw an exception if creating the `urdf` module fails, i.e. `RaveCreateModule` returns `None`.

File: src/prpy/serialization.py [hunk 22]
```diff
@@ -320,44 +354,83 @@ def deserialize_environment(data, env=None, purge=False, reuse_bodies=None):
...
+    
+    if data.get('uri_only', False):
+        
+        if data['is_robot']:
+
+            parts = data['uri'].split()
+            if len(parts)==2 and parts[0].endswith('.urdf') and parts[1].endswith('.srdf'):
+                robot_name = urdf_module_getter().SendCommand('Load {}'.format(data['uri']))
+                if robot_name != name_desired:
+                    raise RuntimeError('error, or_urdf name mismatch!')
```
Comment (mkoval) [inline src/prpy/serialization.py]: This check is incorrect. `or_urdf` loads the robot with a default name generated from the URDF file. This code should call `kinbody.SetName(name_desired)` to override the name with the serialized value.

File: src/prpy/serialization.py [hunk 23]
```diff
@@ -320,44 +354,83 @@ def deserialize_environment(data, env=None, purge=False, reuse_bodies=None):
...
+                kinbody = env.ReadRobotXMLFile(data['uri'])
+                kinbody.SetName(name_desired)
+                env.Add(kinbody, anonymous)
+            
+        else:
+
+            if data['uri'].endswith('.urdf'):
+                kinbody_name = urdf_module_getter().SendCommand('Load {}'.format(data['uri']))
+                if kinbody_name != name_desired:
+                    raise RuntimeError('error, or_urdf name mismatch!')
```
Comment (mkoval) [inline src/prpy/serialization.py]: See above. This logic is incorrect.

File: src/prpy/serialization.py [hunk 24]
```diff
@@ -320,44 +354,83 @@ def deserialize_environment(data, env=None, purge=False, reuse_bodies=None):
...
+                data['uri']
+            )
+        
+        else:
+            
+            kinbody = RaveCreateKinBody(env, '')
+            kinbody.Init(link_infos, joint_infos, data['uri'])
+        
+        kinbody.SetName(name_desired)
+        env.Add(kinbody, anonymous)
```
Comment (mkoval) [inline src/prpy/serialization.py]: All of this code appears to be mistakenly in the `if uri_only` block.

File: src/prpy/serialization.py [hunk 25]
```diff
@@ -381,10 +454,19 @@ def deserialize_kinbody_state(body, data):
...
-        data['dof_branches']
-    )
+    link_transforms = data.get('link_transforms')
+    dof_branches = data.get('dof_branches')
+    if link_transforms is not None and dof_branches is not None:
+        body.SetLinkTransformations(
+            map(deserialize_transform, link_transforms),
+            dof_branches
+        )
+
```
Comment (mkoval) [inline src/prpy/serialization.py]: Nit: Remove newline.

File: src/prpy/serialization.py [hunk 26]
```diff
@@ -381,10 +454,19 @@ def deserialize_kinbody_state(body, data):
...
+    else:
+        deserialization_logger.warn(
+            ('KinBody "{}" does not have link_transforms/dof_branches'
+            + 'saved; falling back to dof_values').format(body.GetName()))
```
Comment (mkoval) [inline src/prpy/serialization.py]: This awkward string concatenation is not necessary. Python does string concatenation like C (your favorite, @cdellin!):
```
>>> print ('banana'
           'orange')
bananaorange

**Comment by mkoval** [comment]
Added another round of comments.

File: src/prpy/planning/logged.py (hunk 12 above)
Comment (cdellin) [inline src/prpy/planning/logged.py]: Added.

File: src/prpy/planning/logged.py (hunk 13 above)
Comment (cdellin) [inline src/prpy/planning/logged.py]: Done! No wicked dictionary comprehension though ... sounds hard to read. (Feel free to try it if you want!)

File: src/prpy/planning/logged.py (hunk 14 above)
Comment (cdellin) [inline src/prpy/planning/logged.py]: Added a comment; this will through a serialization error on failure, which is I think what we want.

File: src/prpy/planning/logged.py (hunk 15 above)
Comment (cdellin) [inline src/prpy/planning/logged.py]: Done!

File: src/prpy/serialization.py (hunk 16 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Done.

File: src/prpy/serialization.py (hunk 17 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Good idea! Done.

File: src/prpy/serialization.py (hunk 18 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Fixed!

File: src/prpy/serialization.py (hunk 19 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Yuppers!

File: src/prpy/serialization.py (hunk 20 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Missing no longer. (Is `UnitaryMemoizer` not sufficiently self-describing? `d-:`)

File: src/prpy/serialization.py (hunk 21 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Done (see later on use).

File: src/prpy/serialization.py (hunk 22 above)
Comment (cdellin) [inline src/prpy/serialization.py]: I think I fixed this ... (I need to remove from environment to change name, no?)

File: src/prpy/serialization.py (hunk 23 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Ditto.

File: src/prpy/serialization.py (hunk 24 above)
Comment (cdellin) [inline src/prpy/serialization.py]: I looked at this, and I think it's correct ... there's an `else:` above, so this code should only be executed if `uri_only` is not set.

File: src/prpy/serialization.py (hunk 25 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Newline obliterated.

File: src/prpy/serialization.py (hunk 26 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Fancy! Done and done.

**Comment by cdellin** [comment]
OK, took care of everything (besides the non-`uri_only` kinbody loading logic which I contend is correct).  I'm ready for round 3!

File: src/prpy/planning/logged.py [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+        # retrieve enviroment
+        if 'robot' in kw_args:
+            env = kw_args['robot'].GetEnv()
+        elif len(args) > 0:
+            env = args[0].GetEnv()
+        else:
+            raise RuntimeError('could not retrieve env from planning request!')
+        
+        # serialize environment
+        envdict = prpy.serialization.serialize_environment(env, uri_only=True)
```
Comment (mkoval) [inline src/prpy/planning/logged.py]: Can you make `uri_only` a constructor parameter so it's not hardcoded to be `True`?

File: src/prpy/planning/logged.py [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+            return traj
+            
+        except prpy.planning.PlanningError as ex:
+            resdict['ok'] = False
+            resdict['exception'] = str(ex)
+            raise
+        
+        except:
+            resdict['ok'] = False
+            resdict['exception'] = 'unknown exception type'
```
Comment (mkoval) [inline src/prpy/planning/logged.py]: You can get the exception here using `str(sys.exc_info[1])` and eliminate the `PlanningError` case.

File: src/prpy/serialization.py [hunk 29]
```diff
@@ -99,30 +103,51 @@ def serialize_environment_file(env, path, writer=None):
...
-    all_joints.extend(body.GetJoints())
-    all_joints.extend(body.GetPassiveJoints())
+def serialize_kinbody(body, uri_only=False):
 
+    uri = body.GetXMLFilename()
+    
+    if uri_only and not uri:
+        serialization_logger.warn(
+            'uri_only passed, but KinBody "{}"\'s GetXMLFilename()'
+            'returns an empty URI.'.format(
```
Comment (mkoval) [inline src/prpy/serialization.py]: Missing space before `returns`.

File: src/prpy/serialization.py [hunk 30]
```diff
@@ -381,10 +475,18 @@ def deserialize_kinbody_state(body, data):
...
+    dof_branches = data.get('dof_branches')
+    if link_transforms is not None and dof_branches is not None:
+        body.SetLinkTransformations(
+            map(deserialize_transform, link_transforms),
+            dof_branches
+        )
+    else:
+        deserialization_logger.warn(
+            'KinBody "{}" does not have link_transforms/dof_branches'
+            'saved; falling back to dof_values'.format(body.GetName()))
```
Comment (mkoval) [inline src/prpy/serialization.py]: Missing a space before `saved`.

File: src/prpy/planning/logged.py (hunk 27 above)
Comment (cdellin) [inline src/prpy/planning/logged.py]: Added!

File: src/prpy/planning/logged.py (hunk 28 above)
Comment (cdellin) [inline src/prpy/planning/logged.py]: I did the former, but I prefer if non-expected exceptions have the additional "Unknown exception type" text.

File: src/prpy/serialization.py (hunk 29 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Done!

File: src/prpy/serialization.py (hunk 30 above)
Comment (cdellin) [inline src/prpy/serialization.py]: Done!

**Comment by cdellin** [comment]
Comments addressed!
